internal/wasm/memory.go

```diff
@@ -193,17 +195,25 @@ func MemoryPagesToBytesNum(pages uint32) (bytesNum uint64) {
 //
 // Returns -1 if the operation resulted in exceeding the maximum memory pages.
 // Otherwise, returns the prior memory size after growing the memory buffer.
-func (m *MemoryInstance) Grow(_ context.Context, newPages uint32) (result uint32) {
+func (m *MemoryInstance) Grow(_ context.Context, delta uint32) (result uint32) {
 	// Note: If you use the context.Context param, don't forget to coerce nil to context.Background()!
 
 	currentPages := memoryBytesNumToPages(uint64(len(m.Buffer)))
+	if delta == 0 {
+		return currentPages
+	}
 
 	// If exceeds the max of memory size, we push -1 according to the spec.
-	if currentPages+newPages > m.Max {
+	newPages := currentPages + delta
+	if newPages > m.Max {
 		return 0xffffffff // = -1 in signed 32-bit integer.
-	} else {
-		// Otherwise, grow the memory.
-		m.Buffer = append(m.Buffer, make([]byte, MemoryPagesToBytesNum(newPages))...)
+	} else if newPages > m.Cap { // grow the memory.
+		m.Buffer = append(m.Buffer, make([]byte, MemoryPagesToBytesNum(delta))...)
+		m.Cap = newPages
+		return currentPages
+	} else { // We already have the capacity we need.
```

Comment:
this is the main optimization as it is the only way I know of to reset the length of a slice w/o calling append or copy